torch/autograd/grad_mode.py

```diff
@@ -42,7 +42,7 @@ def __exit__(self, *args):
     def __call__(self, func):
         @functools.wraps(func)
         def decorate_no_grad(*args, **kwargs):
-            with self:
+            with torch.no_grad():
```

Comment:
I am very confused. Why did `with self` not work?

Reply:
Yeah, I don't understand either, but here is the difference I saw:

Using `with self`:
```
class Decor():
    def __init__(self):
        print('__init__ called')
        self.init_var = 0
        
    def __enter__(self):
        print('__enter__ called')
        return self
    
    def __exit__(self, exc_type, exc_value, exc_traceback):
        print('__exit__ called')
    
    def __call__(self, func):
        def decorate(*args, **kwargs):
            with self:
                return func(*args, **kwargs)
        return decorate

@Decor()
def foo():
    print('inside with statement body')
print("==============")
foo()
######################
__init__ called
==============
__enter__ called
inside with statement body
__exit__ called
```

Using `with Decor()`:
```
class Decor():
    def __init__(self):
        print('__init__ called')
        self.init_var = 0
        
    def __enter__(self):
        print('__enter__ called')
        return self
    
    def __exit__(self, exc_type, exc_value, exc_traceback):
        print('__exit__ called')
    
    def __call__(self, func):
        def decorate(*args, **kwargs):
            with self:
                return func(*args, **kwargs)
        return decorate

@Decor()
def foo():
    print('inside with statement body')
print("==============")
foo()
###################
__init__ called
==============
__init__ called
__enter__ called
inside with statement body
__exit__ called
```

Notice `__init__` gets invoked once more when `with Decor()` decorated function is called

Reply:
Yes this seems expected, because when you do `with Decor()`, a new instance is created. I will think more why `with self` doesn't work.

Reply:
is it sufficient if we know that `self` only represents the class object, and thus with `with self:` context only gets recorded once when decorator applied on a function, where if we use `with DecoratorName()`, a new instance gets created whenever the decorated function is called, and so the latest context will be recorded.

Reply:
Okay I understand now. Recall the original bahavior:
```
import torch

@torch.no_grad()
def nothing(x):
    return x

testin = torch.Tensor([0])
with torch.no_grad():
    print(torch.is_grad_enabled()) # False
    testout = nothing(testin)
    print(torch.is_grad_enabled()) # True
```

This happens, because we save `self.prev = torch.is_grad_enabled()` in `__init__` not in `__enter__`. So when the `no_grad` object is created in `@torch.no_grad()` , it saves `self.prev = True`. And then after `nothing`, the `grad_mode` is restored to `True` at its `__exit__`. To fix this. We should just make `__init__` a no-op, and save the `self.prev` in `__enter__`.

Reply:
Thanks! I will change it accordingly